site/content/resources/adopters.md

```diff
@@ -51,6 +54,22 @@ DaoCloud is an innovation leader in the cloud-native field. With the competitive
 
 DaoCloud build Next Generation Microservices Gateway based on Contour, and also contribute in Contour Community deeply.
 
+**[SnappCloud](https://snapp.ir)**
+
+SnappCloud has developed several solutions to provide a complete self-service and multi-tenant API-GW solution with Contour:
+
+1. [Cerberus](https://github.com/snapp-incubator/Cerberus): Cerberos is a powerful authorization server designed to seamlessly integrate with Contour by implementing the auth_ext interface of Envoy. In the world of modern application deployment and microservices architecture, ensuring secure and controlled access to services is paramount. Cerberos fills this role by providing a dynamic and flexible access control solution tailored to the unique demands of Contour-based applications.
+
+2. [Contour Global Rate Limit Operator](https://github.com/snapp-incubator/contour-global-ratelimit-operator): This project provides a Kubernetes operator that allows users to configure global ratelimits in their HTTPProxy and it configures a RLS service based on [envoyproxy/ratelimit](https://github.com/envoyproxy/ratelimit).
+
+3. [Contour Admission Webhook](https://github.com/snapp-incubator/contour-admission-webhook): This webhook facilitates the validation and mutation of Contour's HTTPProxy resources, ensuring configurations adhere to defined policies and standards. For example, it blocks creation of HTTPProxies with conflicting FQDNs, to prevent a user to invalidate other HTTPProxies in other namespaces.
```

Comment:
@projectcontour/maintainers we are also  finding the need to build this.

Should we bring in an implementation of this into `projecontour` organization? Should we open an issue for that?

Reply:
🤔 we have some examples of using Gatekeeper for this -- see https://projectcontour.io/guides/gatekeeper/ and https://github.com/projectcontour/contour/tree/main/examples/gatekeeper. I think we would also prefer to make more use of CEL validation where possible, since they don't require deploying an additional component.

That said, if there is still a need for a custom admission webhook and is something that there is shared community interest in, we could definitely look at creating a repo for it in the projectcontour org.

Reply:
TIL, that is great

Reply:
let me digest that but seems really useful!

Reply:
Regarding CEL, AFAIK, Gatekeeper and CEL validation in CRDs are generally limited to the properties of the resource itself. So for HTTPProxy host FQDN conflicts, it's not possible to do so. It is possible to do it with other policy engines such as kyverno, as they support [apiCall](https://kyverno.io/docs/writing-policies/external-data-sources/#variables-from-kubernetes-api-server-calls), however it still requires to get all HTTPProxies in all namespaces with every update on HTTPProxies. We implemented a custom webhook, so we can cache all FQDNs to avoid such heavy lookups for every operation.

Reply:
JFYI:
We're pleased to inform you about the release of the new version of our custom admission webhook. This update focuses on improving stability, making it a dependable choice for production environments. The admission webhook maintains its commitment to enforcing security standards for Contour in multi-tenant clusters.